cmake-tool/helpers/cpio.cmake

```diff
@@ -39,48 +39,100 @@ function(MakeCPIO output_name input_files)
     if(NOT "${MAKE_CPIO_UNPARSED_ARGUMENTS}" STREQUAL "")
         message(FATAL_ERROR "Unknown arguments to MakeCPIO")
     endif()
+
     set(archive_symbol "_cpio_archive")
     if(NOT "${MAKE_CPIO_CPIO_SYMBOL}" STREQUAL "")
         set(archive_symbol ${MAKE_CPIO_CPIO_SYMBOL})
     endif()
-    # Check that the reproducible flag is available. Don't use it if it isn't.
+
+    set(cpio_archive "${output_name}.cpio")
+
+    # CMake wants absolute paths when calling file(WRITE ...)
+    get_filename_component(
+        output_name_abs
+        "${output_name}"
+        ABSOLUTE
+        BASE_DIR
+        "${CMAKE_CURRENT_BINARY_DIR}"
+    )
+
+    # Create a script that prepares CPIO archive contents in a tmp folder and
+    # then builds the archive. Some cpio versions support the paramter
+    # "--reproducible" to create the archive with consistent inodes and device
+    # numbering. In addition, we set each file's the 'modified time' to the
+    # epoch (ie 0, but simply using 'touch -d @0' is a GNU extension of the
+    # POSIX standard), and the user/group values to 0:0.
+    # Since some application access the archive contents by index and not by
+    # name, the files must be put into the archive in exactly the same order as
+    # they are listen in 'input_files'. Using simply "ls | cpio ...." instead of
+    # "printf '%s\\n' \${@##*/} | cpio ..." does not guarantee preserving the
+    # order.
     CheckCPIOArgument(cpio_reproducible_flag "--reproducible")
-    set(
-        commands
-        "bash;-c;cpio ${cpio_reproducible_flag} --quiet --create -H newc --file=${CMAKE_CURRENT_BINARY_DIR}/archive.${output_name}.cpio;&&"
+    set(cpio_archive_creator "${output_name_abs}.sh")
+    file(
+        WRITE
+            "${cpio_archive_creator}"
+            # content
+            "#!/bin/sh\n"
+            "# auto-generated file from MakeCPIO(), do not edit\n"
```

Comment:
Something like `set -e` is needed here or when the script is invoked to handle when any of the inner commands return an error.

Reply:
added, thanks.